### ISSUE TITLE
Documentation clarification for universal link settings

### ISSUE THREAD

**Issue by barocsi** [BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK#132]
Hello there is a slight ambiguity in documentation files,
this link
https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK
and this link
https://start.branch.io/?__hssc=&__hstc=196949811.01a59d63dfb29d34d9136d1e2fa52ef4.1453236561358.1461623127530.1461671981260.17&__hsfp=1189881812&hsCtaTracking=d8a94b5f-4160-45df-9c13-c71f7c5e8031%7Cd4846147-2f7f-4600-ab4d-9facf5fb409b#/integrate-sdk/steps
says

> iOS: Enable Universal Links
> 
> In iOS 9.2, Apple dropped support for URI scheme redirects. You must enable Universal Links if you want Branch-generated links to work in your iOS app. To do this:
> 
> Enable Associated Domains capability on the Apple Developer portal when you create your app's bundle identifier.
> In your Dashboard Link Settings, tick the Enable Universal Links checkbox and provide the Bundle Identifier and Apple Team ID in the appropriate boxes.
> Add the following entry to your application's config.xml:
> 
```
> <branch-config>
>     <ios-team-id value="your_ios_team_id" />
>     <host name="bnc.lt" scheme="https" />
> </branch-config>
```
> 

But this link 
https://dev.branch.io/getting-started/sdk-integration-guide/guide/cordova/
says that

> Caution: You should also configure Universal Links
> 
> In iOS 9.2, Apple dropped support for URL scheme redirects. You must also enable Universal Links to get full functionality from Branch on iOS 9.2+.

Where it says that I must enable universal links, ok:
https://dev.branch.io/getting-started/universal-app-links/
Says on cordova guide tab, that 

> Prerequisite: Complete the SDK Integration Guide first
> 
> This guide requires you to have already integrated the Branch SDK into your app.

Where it says to go back to
https://dev.branch.io/getting-started/sdk-integration-guide/guide/cordova/

Is this some sort of circular dependency?

But okay, I try to follow the enabling universal links part on
https://dev.branch.io/getting-started/sdk-integration-guide/guide/cordova

I have an app id, associated domains enabled, universal links enabled on Branch dashboard, and get to the APP ENTITLEMENT PART where it says
>
```
> <branch-config>
>     <ios-team-id value="your_ios_team_id" />
>     <host name="xxxx.app.link" scheme="https" />
>     <host name="xxxx-alternate.app.link" scheme="https" />
> </branch-config>
```
>

Now, this is interesting, because on the other links it said:
>
```
> <branch-config>
>     <ios-team-id value="your_ios_team_id" />
>     <host name="bnc.lt" scheme="https" />
> </branch-config>
```
>

So. What to do next? Which one is right and which one is not. There are referrings, that if I have a bnc.lt on my branch dashboard, I should use bnc.lt. Ok, I keep it that way.

In iOS 8.4 it works, however first opens in Safari.
In iOS 9.2+ it just does not work. It opens the deepview in safari and stays there.

I keep trying to discover more issues in https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/issues/129#issuecomment-214938375

**Comment by renesansz** [comment]
Hey @barocsi ,
I'm not sure if the documentation in `https://dev.branch.io` is updated.
However, the way the testbed has been implemented follows on what you see in the `README.md`.
I do only have tested this in the emulator and so far it didn't also opened the app.
Let me investigate more in the issue and what's the probable cause.

**Comment by barocsi** [comment]
Thanks,
Also not that in iOS 9.3.1 also does not work with live device. (mentioned in https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/issues/129#issuecomment-214938375).

**Comment by barocsi** [comment]
I have noticed something.
1. Open the app
2. Open Facebook or Gmail, click on a bnc.lt link
3. Opens in safari, but dont opens in the app
4. Close the app
5. Start the app
6. It opens the originally intended url from step 2.

**Comment by barocsi** [comment]
ok I have found something very similar here:

http://stackoverflow.com/questions/35314966/wkwebview-still-not-handling-universal-link

**Comment by barocsi** [comment]
And a possible another explanation:
http://stackoverflow.com/questions/35172678/universal-links-not-working-with-wkwebview

**Comment by barocsi** [comment]
Another bit of useful information:
https://github.com/EddyVerbruggen/Custom-URL-scheme/issues/144

**Comment by renesansz** [comment]
@barocsi , thanks for the leads.
I'll replicate the issue first and see what I can found and submit a fix.

**Comment by barocsi** [comment]
However if I turn off universal link option on dashboard, should it employ javascript method?
According to this 
https://github.com/EddyVerbruggen/Custom-URL-scheme
There some sort of JS to should handle app opening on my branch deepview page, but it still only works below iOS 9.2.
> As of iOS 9.2, the dialog Open in "mycoolapp"? no longer blocks JS, so if you have a short timeout that opens the app store, the user will be taken to the store before they have a chance to see and answer the dialog. See below for available solutions.

**Comment by renesansz** [comment]
hmm, not sure if it does need to employ a javascript method for this.
This plugin was based on the previous version of the plugin, so not sure if this is still applicable for the latest. 😕

**Comment by barocsi** [comment]
http://stackoverflow.com/questions/35138257/branch-deep-link-not-opening-application

This could be an issue.

**Comment by aaustin** [comment]
Hey @barocsi - Can you review the support section of the documentation? https://dev.branch.io/getting-started/universal-app-links/support/ios/#appsbrowsers-that-support-universal-links In order to test, we recommend using messages or notes on a live device.

This is likely not a problem with the plugin, as we have successfully tested Universal Link with this version.

**Comment by davebritt** [comment]
I am having the exact same issue as described by barocsi above  ... I have reviewed the troubleshoting guide and doco as suggested. I am testing from the iOS email client on a live device  (which is what we need to work in production).

**Comment by barocsi** [comment]
# Wrap up - iOS 9.3

_Disclaimer: this is just external testing experience mixed with documentation, not backed up by any developer from Branch._

## So far the following is discovered with setting up universal links:

1. You must match team ids. This means if you run on a device with a code signing profile that does not match the id you have entered in branch.io dashboard it will not work.

2. For me it only worked with a device with AD-HOC distribution mode. On simulator, Universal Link does not really work according to [this](https://forums.developer.apple.com/thread/44526)

3. If you change branch.io universal link mode to ON, the links api starts generating for you new type of links from that point and will change from legacy bnc.lt/xyz format to bnc.lt/4char_id/xyz, which is required by  [Apple App Site Association file format](https://developer.apple.com/library/ios/documentation/General/Conceptual/AppSearch/UniversalLinks.html). This means if you switched to universal link format and think your legacy link will change, it wont. You must test with newly generated links. Legacy links only work below on certain iOS versions because of the point 4. below. It might be considered for Branch team to add some sort of redirect for legacy links to add cross compatibility between links if they do not fix 9.3 URL Scheme fallback. Or just make it clear in the docs.

4. There are [limitations](https://dev.branch.io/getting-started/universal-app-links/support/ios/#appsbrowsers-that-support-universal-links) of universal links, sometimes they open, sometimes they dont, there is a nice list of applications supporting or not supporting universal links. For example, I can tap on an universal link in GMail and it opens in my app, but in Facebook it stays there.

5. If you messed around with opening a link in your app, there usually is a popup. It unfortunately remembers to it, and you REALLY need to discover what is the saved behavior. For example in Safari, you see a deepview instead an app opening, just pull the page down and you see a line with a button to open it in the app. In facebook there is a settings button on the bottom right as well.

> When a user taps a universal link that you handle, iOS also examines the user’s recent choices to determine whether to open your app or your website. For example, a user who has tapped a universal link to open your app can later choose to open your website in Safari by tapping a breadcrumb button in the status bar. After the user makes this choice, iOS continues to open your website in Safari until the user chooses to open your app by tapping OPEN in the Smart App Banner on the webpage.

 
## Conclusion

Universal Link does not always trigger app opening. URL Scheme could be still harnessed if possible. As #129 discovers Branch.io displays different deepviews depending on your iOS version meaning if you visit with iOS 8.4 you get an iframe src hack opening yourapp:// . However in iOS 9.2+ it stopped working, so it was discontinued in favor of Universal Links. 
But as #129 reference speculates it could be still there using [some sort of javascript logic](https://blog.branch.io/ios-9.2-redirection-update-uri-scheme-and-universal-links) so in case Universal Link does not kick in within a given timeframe, URL Scheme would, as st.derrick mentioned in [this SO thread](http://stackoverflow.com/questions/32689483/ios9-try-to-open-app-via-scheme-if-possible-or-redirect-to-app-store-otherwise):

The iframe trick no longer works -- my guess is that Apple knows it will encourage more developers to implement Universal Links, more quickly.

> You can still set window.location='your-uri-scheme://'; and fallback to the App Store after 500ms. There is a "dance" between popups if you take this approach, as we do at Branch (we do as a fallback if Universal Links don't work).

```javascript
window.location = 'your-uri-scheme://'; // will result in error message if app not installed
setTimeout(function() {
   // Link to the App Store should go here -- only fires if deep link fails                
   window.location = "https://itunes.apple.com/us/app/myapp/id123456789?ls=1&mt=8";
}, 500);
```

My question is: why this piece of code is missing from bnc.lt/4char/xy links when opened from iOS9.3 as mentioned in #129 ?

**Comment by aaustin** [comment]
Unfortunately, Facebook still blocks both Universal Links *and* URI scheme approach so you can never test with that browser. Additionally, as you've found, Universal Links don't work on the simulator or on most browsers. The best way to test is with Notes or Messages on a live device.

The reason we don't include the fallback URI scheme method on bnc.lt/4char/xy links on 9.3 is that 100% of the time, the URI scheme popup is ignored before redirecting to the store. Since Apple made that a non-blocking modal, Javascript continues to execute and the redirect occurs before the user can respond. The result is a jarring, modal flash and redirect experience which ultimately results in the user ending up on the App Store page 100% of the time. Therefore, we decided to make that redirect a lot cleaner and faster by just issuing a 3XX redirect to the App Store. We do, however, use this URI scheme redirect technique on Gmail iOS 9+ since that still works.

Ultimately, if you enable Universal Links and support URI, the majority of cases on iOS are handled well. The only one that is actually broken is Facebook, and it's broken for everyone unfortunately since they block both techniques.

I hope this helps clear things up. Closing the ticket for now.
